plugin-server/src/main/ingestion-queues/session-recording/services/overflow-detection.ts

```diff
@@ -0,0 +1,45 @@
+import LRUCache from 'lru-cache'
+import { Gauge } from 'prom-client'
+
+import { Limiter } from '../../../../utils/token-bucket'
+
+export enum OverflowState {
+    Okay,
+    Triggered, // Recently triggered the overflow detection
+    Cooldown, // Already triggered the overflow detection earlier than cooldownSeconds
+}
+
+export const overflowTriggeredGauge = new Gauge({
+    name: 'overflow_detection_triggered_total',
+    help: 'Number of entities that triggered overflow detection.',
+})
+
+/**
+ * OverflowDetection handles consumer-side detection of hot partitions by
+ * accounting for data volumes per entity (a session_id, a distinct_id...).
+ *
+ * The first time that the observed spike crosses the thresholds set via burstCapacity
+ * and replenishRate, observe returns Triggered. Subsequent calls will return Cooldown
+ * until cooldownSeconds is reached.
```

Comment:
Just out of curiosity, any plans for `Triggered` vs `Cooldown`? Presumably they will be spamming events at us, and so only one caller will see `Triggered`, and I'm not sure what it would be used for that's different from `Cooldown`.

Reply:
`Triggered` will cause the `ZADD` to be called, `Cooldown` will just skip it to reduce load on Redis and iowaits.